AWSPinpoint/AWSPinpointEndpointProfile.m

```diff
@@ -121,12 +126,32 @@ - (BOOL) isApplicationLevelOptOut:(AWSPinpointContext *) context {
     return NO;
 }
 
+- (void) setOptOut:(NSString *)optOut {
+    NSString *overrideDefaultOptOut = [self.userDefaults stringForKey:AWSPinpointOverrideDefaultOptOutKey];
+    if (![overrideDefaultOptOut isEqualToString:optOut]) {
+        [self.userDefaults setObject:optOut forKey:AWSPinpointOverrideDefaultOptOutKey];
+        [self.userDefaults synchronize];
+    }
+
+    _optOut = optOut;
+}
+- (NSString *)optOut {
+    return _optOut;
+}
+
 - (void) setEndpointOptOut:(BOOL) applicationLevelOptOut {
-    BOOL isUsingPinpointForNotifications = [AWSPinpointNotificationManager isNotificationEnabled] && [self.address length];
-    BOOL isOptedOutForNotifications = !isUsingPinpointForNotifications;
+    NSString *overrideDefaultOptOut = [self.userDefaults stringForKey:AWSPinpointOverrideDefaultOptOutKey];
+    if ([overrideDefaultOptOut length]) {
+        @synchronized (self) {
+            self->_optOut = overrideDefaultOptOut;
+        }
```

Comment:
won't this result in the value being `"NONE"` even when the notifications are disabled at an app level or the `self.address` is missing?

Reply:
I don't think i fully understand the scenario ur talking about, but if the developer sets the value to `"NONE"`, it will persist in the user defaults, so it will always update the _optOut to the value that it was set in the user defaults, in this case we respect what is there and set it to `"NONE"`. 

if we look at a scenario where the optOut is not explicitly set by the developer, it will take on the value of ALL or NONE as before. NONE only when both UIApplication.isRemoteNotificationEnabled && devicetoken/address is set. otherwise set it to ALL.

If the developer builds a toggle that sets it to ALL or NONE (or anything for that matter), this will be immediately persisted in user defaults, and always take precedent over the default behavior

Reply:
so the scenario is:
- developer sets to `"NONE"`, `overrideDefaultOptOut ` is now `"NONE"`
- `address` is `nil`
- `optOut` is submitted as `"NONE"`
- *"missing address error"*

Reply:
i think service response is out of our control. but we can avoid it by doing `if overrideDefaultOptOut == "NONE" and _address == nii, set it to ALL`

Reply:
I think the previous logic needs little modification to work with the newly introduced user override:

```objc
// line 129 was:
self->_optOut = (applicationLevelOptOut || isOptedOutForNotifications)? @"ALL": @"NONE";

// it needs to change to
self->_optOut = (applicationLevelOptOut || isOptedOutForNotifications)? @"ALL": USER_SET_VALUE_OR_NONE_BY_DEFAULT;
```

Reply:
thanks, updating this and added unit test for setting to `None` but one of the other is not set, will result in `ALL`